### PR TITLE
[build] [dune] Add world-prelude target to produce a subset of world

### PR DESCRIPTION
**Pull request by ejgallego**
Some users want to have a subset of the world target for faster
interaction with non-dunerized builds.

This is not seamless done in Dune due to install layouts not composing
in general [so implementing `%{pkg:coq:foo}` is hard when the `coq`
package is not in the workspace]

Instead, we provide a convenience target to build some subsets of
world until we improve support in dune.

Likely, Dune should support the `@install` alias scoped to
sub-directories but there seems to be a problem with it as of today.

### CODE REVIEW

**Comment by ejgallego** [comment]
cc: @ppedrot

File: Makefile.dune
Comment (SkySkimmer) [inline Makefile.dune:57]:
```suggestion
	@echo "  - world-states: build Coq's prelude and setup a functional layout on _build/install"
```

File: Makefile.dune
Comment (SkySkimmer) [inline Makefile.dune:171]:
```suggestion
PRELUDE_FILES=$(wildcard theories/Init/*.v)
```

**Comment by SkySkimmer** [comment]
Why not replace the states target?

**Comment by ejgallego** [comment]
Thanks for feedback @SkySkimmer , updated.

> Why not replace the states target?

As of today it is way more fine grained as it doesn't build coq-core.install , I dunno.

I took this `world-lib` scheme as maybe we want some other such as `world-arith` `world-list` etc...

**Comment by SkySkimmer** [comment]
Why wouldn't we call them `arith` `list` etc?

**Comment by ejgallego** [comment]
> Why wouldn't we call them `arith` `list` etc?

I kinda like to have some structure on targets, but no strong preference here; but I find useful to type `make world-` and use shell completion for example.

**Comment by ejgallego** [comment]
Thanks @SkySkimmer , comments fixed.

**Comment by Zimmi48** [comment]
As someone who has been using Dune for months, I never use the `states` target from `Makefile.dune`, always call `dune build` or `dune exec` on one of the shims directly. Therefore, given that `states` is mostly for convenience when migrating from the legacy build system, I would recommend reusing the name `states` for this new target instead of inventing a new one.

**Comment by ejgallego** [comment]
> As someone who has been using Dune for months, I never use the `states` target from `Makefile.dune`, always call `dune build` or `dune exec` on one of the shims directly. Therefore, given that `states` is mostly for convenience when migrating from the legacy build system, I would recommend reusing the name `states` for this new target instead of inventing a new one.

I dunno, the new target is actually far from what the old states did, the current states does exactly what the old states did, and indeed it didn't setup a lot of stuff including coq_makefile.

Personally, I always found the name `states` for a target really puzzling, when I was a begginer in Coq I could not decipher what was and it really surprised me. So if something, I would just remove the states targets from dune at all as it is confusing.

I am still convinced that subsets of the world target should be named `world-foo`, both for discoverability and meaning.

So indeed I have renamed the new target to `world-prelude` which is IMHO more accurate.

**Comment by SkySkimmer** [comment]
Isn't everything a subset of world?

**Comment by ejgallego** [comment]
> Isn't everything a subset of world?

The bottom element of the world-lattice should be the minimally functioning public install layout, so empty install for example is not an element of the lattice.

These targets are anyways moot once we dunerize more ci devs.

**Comment by ppedrot** [comment]
> the current states does exactly what the old states

No. The goal of the `states` target was to get a minimally functioning `coqc` / `coqtop`. I am afraid that currently, the binaries produced this way cannot be considered functional. (Historically, `states` was called this way because Coq relied on explicitly marshalling its current state à la `-inputstate`, and this target was generating a form of bootstrap.)

> These targets are anyways moot once we dunerize more ci devs.

I don't agree. I often need to fire up a minimally functioning Coq to draft an idea / write some boilerplate. In the current state, I cannot do that without having to go through the pain of compiling 15 minutes of irrelevant stuff. I know I can go through the shim, but it's not going to work as soon as I have a non-trivial scratchpad development (i.e. if it requires a module to compile, because otherwise I need to write a full-blown library). If this can be perceived as an argument that the stdlib should be somehow considered a separate package, so be it, but please let me have a minimal working Coq binary that I can compile in less than 2 minutes.

Regarding the name of the target, I don't have a strong opinion, but 1. the current `states` target is never used by dune power users and 2. it's useless as a helper targer for legacy-inclined people, so it doesn't seem far-fetched to me to *carrément* replace the current `states` target with something that does the job. Otherwise I'd never use the other `states` target.

**Comment by Zimmi48** [comment]
I'm quite convinced by the arguments you gave for `world-*` targets so let's go with this if no one opposes. However, let's also remove the `states` target which is just too confusing for developers migrating. If it's not too hard to do, I would also provide `world-arith` now. I think it would be useful pretty often.

**Review by Zimmi48 — APPROVED** [review]

**Comment by ejgallego** [comment]
> I'm quite convinced by the arguments you gave for `world-*` targets so let's go with this if no one opposes. However, let's also remove the `states` target which is just too confusing for developers migrating. If it's not too hard to do, I would also provide `world-arith` now. I think it would be useful pretty often.

Note that `states` will get fixed in #14059 , which is now next in list after we merged the "remove -local" PR, so maybe we want to keep it?

I can add more `world-targets` but it is gonna take a bit as there are more urgent issues right now.

**Comment by ejgallego** [comment]
> I still don't understand the world- prefix.

world-bar , is coq-core + bar, where bar is considered a set of .vo files , world is an alias for `world-all` , coq-core is an alias for `word-zeroset`.

So coq-core is the base of "working coq install", where all the tools, stubs, .cmt / .cmxa files etc are in place [so plugins can link, etc...] , then add some vo files on top of that.

We can use a different naming if you are thinking of something else.

**Comment by Zimmi48** [comment]
> Note that `states` will get fixed in #14059 , which is now next in list after we merged the "remove -local" PR, so maybe we want to keep it?

How does this PR fix `states` and what does that even mean? As of today, `states` does work, it's just not the same meaning as in the legacy build system. Do you mean that it will be the same meaning after this PR? But in this case, what is the difference with `world-prelude`?

**Comment by ejgallego** [comment]
> How does this PR fix states and what does that even mean?

Sorry I was not clear, it means that once you do `make states`, _build/install/default/bin/coqc will correctly find coqlib.

The difference with world-prelude is that it doesn't force install layout to be setup.

**Comment by Zimmi48** [comment]
OK thanks.

I propose to merge this PR tomorrow if there are no more comments / no opposition.

**Comment by Zimmi48** [comment]
@coqbot merge now

**Pull request merged**
